[LV] Update LoopVectorizationPlanner::emitInvalidCostRemarks to handle reduction plans

bababuck opened the pull request:
Fixes #165359

TypeSwitch statement that didn't account for reduction plans. I noticed this comment in the original MR that wasn't resolved, https://github.com/llvm/llvm-project/pull/99322#discussion_r1691509886. I wonder if we want a more permanent solution here (i.e. handle the default case gracefully) rather than adding new classes as they are created.

bababuck committed:
[LV] Add test for failing reduction remark case

Issue #165359.

https://godbolt.org/z/z1P6Wxj7d

fhahn commented on llvm/test/Transforms/LoopVectorize/bug165359.ll:
```diff
@@ -0,0 +1,159 @@
+; ModuleID = 'FIRModule'
+source_filename = "FIRModule"
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
+target triple = "aarch64-unknown-linux-gnu"
```
If the test requires a target-triple, it needs to be moved into the architecture specific sub-dirs, otherwise it will fail when the target is not enabled.

You may be able to reproduce without target-triple by passing something like `-force-vector-width=4`

fhahn commented on llvm/test/Transforms/LoopVectorize/bug165359.ll:
```diff
@@ -0,0 +1,159 @@
+  br label %.preheader, !dbg !30
+
+.preheader:                                       ; preds = %.preheader.preheader, %.preheader
+  %7 = phi double [ %17, %.preheader ], [ %.pre18, %.preheader.preheader ], !dbg !25
```
Would be good to use descriptive names for the various blocks and value names

bababuck committed:
[LV] Update LoopVectorizationPlanner::emitInvalidCostRemarks to handle reduction plans

Issue #165359.

TypeSwitch did not handle the case of retrieving the opcode from a VPRedutionPlan type.

bababuck commented:
Thanks for the feedback. Updates:
- Test moved to AArch64 folder
- Test simplified to minimal example
- Variable names cleaned up for readability.
- Rebased onto more recent `main`

bababuck committed:
[LV] Update bug165359.ll test to demonstrate new output

bababuck committed:
Update test to return result rather than storing

bababuck committed:
Pass in one double rather than loading it

One double still loaded.

bababuck committed:
Remove no alias from remaining pointer

Not needed for test.

bababuck committed:
Switch to canonical loop induction form

bababuck commented:
Thanks for the further feedback, the requested changes have all been pushed. I'll make sure to incorporate that feedback into future tests as well.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/bug165359.ll:
```diff
@@ -0,0 +1,25 @@
+
+loop:
+  %acc0 = phi double [ %fadd0, %loop ], [ %d0, %entry ]
+  %counter = phi i64 [ 0, %entry ], [ %counter_updated, %loop ]
```
%iv for consistency with other tests

fhahn reviewed (approved):
LGTM with a few more suggestions inline, thanks

bababuck committed:
Add newline to end of file

bababuck committed:
Disable output from test since not checked

bababuck committed:
Rename %counter -> %iv

bababuck committed:
Rename %iv_updated -> %iv.next

bababuck committed:
Capture output from pass-remarks-analysis rather than debug-only
